### PR TITLE
Clean up ARM reserved register handling

### PR DESCRIPTION
**Pull request by BruceForstall**
Tighten up and correct the math in compRsvdRegCheck.

### CODE REVIEW

**Comment by BruceForstall** [comment]
@kbaladurin I noticed some things that seemed wrong when reviewing https://github.com/dotnet/coreclr/pull/18031. This change builds on yours and fixes a bunch of (potential) issues. I was able to construct a repro case that failed on Windows ARM32 with the same failure as https://github.com/dotnet/coreclr/issues/17882. I'll look into adding it to the test suite.

**Comment by BruceForstall** [comment]
@dotnet-bot test Ubuntu arm Cross Checked jitstress1 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked jitstress2 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked normal Build and Test
@dotnet-bot test Windows_NT arm Cross Checked jitstress1 Build and Test
@dotnet-bot test Windows_NT arm Cross Checked jitstress2 Build and Test
@dotnet-bot test Windows_NT arm Cross Checked normal Build and Test

**Comment by kbaladurin** [comment]
@BruceForstall thank you. As this patch includes changes #18031 could I close last one?

**Comment by BruceForstall** [comment]
> @BruceForstall thank you. As this patch includes changes #18031 could I close last one?

No, I merged yours, as it is tested now and less impactful. I'll rebase this on top of your change.

**Comment by kbaladurin** [comment]
ok, thank you!

**Comment by BruceForstall** [comment]
@dotnet-bot test Ubuntu arm Cross Checked jitstress1 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked jitstress2 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked normal Build and Test
@dotnet-bot test Windows_NT arm Cross Checked jitstress1 Build and Test
@dotnet-bot test Windows_NT arm Cross Checked jitstress2 Build and Test
@dotnet-bot test Windows_NT arm Cross Checked normal Build and Test

**Comment by BruceForstall** [comment]
test Windows_NT x64 Checked CoreFX Tests
test Windows_NT x64 Release CoreFX Tests

**Comment by BruceForstall** [comment]
Frameworks PMI diffs:

```
Summary:
(Lower is better)
Total bytes of diff: -2896 (-0.01% of base)
    diff is an improvement.
Top file regressions by size (bytes):
          46 : System.Memory.dasm (0.04% of base)
          44 : System.Net.Sockets.dasm (0.02% of base)
          32 : System.Data.Common.dasm (0.00% of base)
          28 : System.Security.Cryptography.Cng.dasm (0.03% of base)
          22 : xunit.runner.utility.dotnet.dasm (0.02% of base)
Top file improvements by size (bytes):
        -684 : Microsoft.CodeAnalysis.VisualBasic.dasm (-0.01% of base)
        -542 : Microsoft.CodeAnalysis.CSharp.dasm (-0.01% of base)
        -214 : System.Collections.dasm (-0.06% of base)
        -196 : System.Private.Xml.dasm (-0.01% of base)
        -160 : System.Collections.Concurrent.dasm (-0.08% of base)
72 total files with size differences (53 improved, 19 regressed), 58 unchanged.
Top method regessions by size (bytes):
          94 : System.Data.Common.dasm - System.Data.FunctionNode:EvalFunction(int,ref,ref,int):ref:this
          58 : Microsoft.VisualBasic.dasm - Microsoft.VisualBasic.CompilerServices.OverloadResolution:CollectOverloadCandidates(ref,ref,int,ref,ref,bool,ref,byref,byref,ref):ref
          36 : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.BoundForToStatement:Update(ref,ref,ref,bool,ref,ref,ref,ref,struct,ref,ref):ref:this
          32 : System.Private.Xml.dasm - System.Xml.Serialization.XmlSchemaImporter:ImportDataType(ref,ref,ref,ref,int,bool):ref:this
          28 : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.InContainerBinder:LookupSymbolsInSingleBinder(ref,ref,int,ref,int,ref,bool,byref):this
Top method improvements by size (bytes):
         -56 : Microsoft.CodeAnalysis.dasm - Parser:ParseNamedTypeSymbol(ref,byref,ref,ref,ref)
         -50 : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.Emit.PEAssemblyBuilderBase:IAssemblyGetFiles(struct):ref:this
         -50 : System.Private.Xml.dasm - System.Xml.Schema.XmlSchemaValidator:ElementValidationError(ref,ref,ref,ref,ref,int,int,ref)
         -48 : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.ExpressionLambdaRewriter:ConvertBooleanOperator(ref):ref:this
         -48 : System.Collections.dasm - System.Collections.Generic.SortedSet`1[Vector`1][System.Numerics.Vector`1[System.Single]]:.ctor(ref,ref):this
1059 total methods with size differences (681 improved, 378 regressed), 190239 unchanged.
35 files had text diffs but not size diffs.
NuGet.Packaging.dasm had 1601 diffs
System.ObjectModel.dasm had 912 diffs
xunit.core.dasm had 598 diffs
System.IO.IsolatedStorage.dasm had 565 diffs
xunit.assert.dasm had 562 diffs
Completed analysis in 169.98s
```

There are many cases where a frame no longer allocates the R10 reserved register.

**Comment by BruceForstall** [comment]
@briansull PTAL
cc @dotnet/jit-contrib

File: src/jit/compiler.cpp [hunk 1]
```diff
@@ -4376,69 +4380,117 @@ bool Compiler::compRsvdRegCheck(FrameLayoutState curState)
...
-    //                -
-    //          high addresses
-
-    // With codeGen->isFramePointerRequired we use R11 to access incoming args with positive offsets
-    // and use R11 to access LclVars with negative offsets in the non funclet or
-    // main region we use SP with positive offsets. The limiting factor in the
-    // codeGen->isFramePointerRequired case is that we need the offset to be less than or equal to 0x7C
-    // for negative offsets, but positive offsets can be imm12 limited by vldr/vstr
-    // using +/-imm8.
+    //          ... low addresses ...
```
Comment (briansull) [inline src/jit/compiler.cpp]: This seems out of place here.   =>` ... low addresses ...`

**Comment by BruceForstall** [comment]
ARM-only change; ignoring x64 corefx failure

**Pull request merged**